### PR TITLE
workspace: Move checkWorkspace

### PR DESCRIPTION
**Pull request by frznlogic**
This commit moves the checkWorkspace method from the SoftwareContainer to the
ContainerUtilityInterface and the usage into the SoftwareContainerAgent class
as the workspace is connected to the agent, not each container. A new check
was added to the SoftwareContainer for the containerRoot (previously
workspace).

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

### CODE REVIEW

**Comment by frznlogic** [comment]
@breakreturn could you make sure you are ok with the changes to the ContainerUtilityInterface. I'm not sure if this is the right way of doing this and thought i'd push it early so you can have some input.

File: agent/src/containerutilityinterface.h [hunk 1]
```diff
@@ -61,15 +65,24 @@ class ContainerUtilityInterfaceError : public SoftwareContainerError
  * This class contains utility functions
  */
 
-class ContainerUtilityInterface {
+class ContainerUtilityInterface :
+    private FileToolkitWithUndo {
```
Comment (breakreturn) [inline agent/src/containerutilityinterface.h]: Put the brace on the next line or put the inheritance stuff on the same line please.

File: agent/component-test/softwarecontaineragent_componenttest.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ class SoftwareContainerAgentTest: public ::testing::Test
 
         try {
             std::shared_ptr<SoftwareContainerFactory> factory = std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface(config));
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_componenttest.cpp]: to be concise could you use somethin like that instead 
`auto utility = std::make_shared<ContainerUtilityInterface>(config);`

Comment (tobsan) [inline agent/component-test/softwarecontaineragent_componenttest.cpp]: I would prefer you only pass the one value you need from the config. If we need more values later on we can change it to take the whole config as an argument.

File: agent/src/containerutilityinterface.cpp [hunk 3]
```diff
@@ -70,4 +78,21 @@ void ContainerUtilityInterface::removeOldContainers()
     delete containerNames;
 }
 
+void ContainerUtilityInterface::checkWorkspace()
+{
+    const std::string rootDir = m_config->getStringValue("SoftwareContainer", "shared-mounts-dir");
+    if (!isDirectory(rootDir)) {
+        log_debug() << "Container root " << rootDir << " does not exist, trying to create";
+        std::unique_ptr<CreateDir> createDirInstance = std::unique_ptr<CreateDir>(new CreateDir());
```
Comment (kursatkobya) [inline agent/src/containerutilityinterface.cpp]: and maybe initiate it like 
`std::unique_ptr<CreateDir> createDirInstance{new CreateDir()}`

File: libsoftwarecontainer/src/container.cpp [hunk 4]
```diff
@@ -190,9 +190,9 @@ bool Container::create()
     m_rootFSPath = buildPath(s_LXCRoot, containerID, "rootfs");
 
     if (m_enableWriteBuffer) {
-        const std::string rootFSPathLower = m_containerRoot + m_id + "/rootfs-lower";
-        const std::string rootFSPathUpper = m_containerRoot + m_id + "/rootfs-upper";
-        const std::string rootFSPathWork  = m_containerRoot + m_id + "/rootfs-work";
+        const std::string rootFSPathLower = m_containerRoot + "/rootfs-lower";
+        const std::string rootFSPathUpper = m_containerRoot + "/rootfs-upper";
+        const std::string rootFSPathWork  = m_containerRoot + "/rootfs-work";
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp:195]: This seems to be unrelated with the checkContainerRoot stuff. Please split it into its own commit.

**Review by tobsan — CHANGES_REQUESTED** [review]

File: agent/src/main.cpp [hunk 5]
```diff
@@ -219,7 +219,7 @@ int main(int argc, char **argv)
         std::shared_ptr<SoftwareContainerFactory> factory =
             std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
         std::shared_ptr<ContainerUtilityInterface> utility =
-            std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface(config));
```
Comment (kursatkobya) [inline agent/src/main.cpp]: again using make_shared would make it more concise and readable

File: libsoftwarecontainer/src/container.cpp (hunk 4 above)
Comment (frznlogic) [reply libsoftwarecontainer/src/container.cpp:195]: It's related, the path available in SoftwareContainer is now "proper", ie it contains the m_id also.

File: agent/src/containerutilityinterface.h (hunk 1 above)
Comment (frznlogic) [reply agent/src/containerutilityinterface.h]: Fixed

File: agent/component-test/softwarecontaineragent_componenttest.cpp (hunk 2 above)
Comment (frznlogic) [reply agent/component-test/softwarecontaineragent_componenttest.cpp]: Fixed for both instances.

File: agent/src/containerutilityinterface.cpp (hunk 3 above)
Comment (frznlogic) [reply agent/src/containerutilityinterface.cpp]: Not sure I understand?

File: agent/component-test/softwarecontaineragent_componenttest.cpp (hunk 2 above)
Comment (frznlogic) [reply agent/component-test/softwarecontaineragent_componenttest.cpp]: I tend to disagree. Or rather, I'm ambivalent as I can see a need for the data inside the class as I interpret the use of it. Would you be terribly upset at keeping it?

File: agent/src/containerutilityinterface.cpp (hunk 3 above)
Comment (frznlogic) [reply agent/src/containerutilityinterface.cpp]: Ok, done!

**Comment by frznlogic** [comment]
I will split the commits apart a bit tomorrow.

File: agent/component-test/softwarecontaineragent_componenttest.cpp [hunk 6]
```diff
@@ -104,8 +104,8 @@ class SoftwareContainerAgentTest: public ::testing::Test
                                                                   ConfigDependencies());
 
         try {
-            std::shared_ptr<SoftwareContainerFactory> factory = std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            auto factory = std::make_shared<SoftwareContainerFactory> ();
+            auto utility = std::make_shared<ContainerUtilityInterface> (config);
```
Comment (tobsan) [inline agent/component-test/softwarecontaineragent_componenttest.cpp]: Remove the spaces before opening parenthesis.

File: agent/src/containerutilityinterface.cpp [hunk 7]
```diff
@@ -70,4 +78,21 @@ void ContainerUtilityInterface::removeOldContainers()
     delete containerNames;
 }
 
+void ContainerUtilityInterface::checkWorkspace()
+{
+    const std::string rootDir = m_config->getStringValue("SoftwareContainer", "shared-mounts-dir");
+    if (!isDirectory(rootDir)) {
+        log_debug() << "Container root " << rootDir << " does not exist, trying to create";
+        std::unique_ptr<CreateDir> createDirInstance{new CreateDir()};
```
Comment (tobsan) [inline agent/src/containerutilityinterface.cpp]: Why the { syntax instead of ( ?

File: agent/src/containerutilityinterface.h [hunk 8]
```diff
@@ -61,15 +65,24 @@ class ContainerUtilityInterfaceError : public SoftwareContainerError
...
  */
 
-class ContainerUtilityInterface {
+class ContainerUtilityInterface : private FileToolkitWithUndo
+{
     LOG_DECLARE_CLASS_CONTEXT("CUI", "Container Utility Interface");
 public:
-    virtual ~ContainerUtilityInterface() {}
+    ContainerUtilityInterface(std::shared_ptr<Config> config);
+    ~ContainerUtilityInterface() {}
```
Comment (tobsan) [inline agent/src/containerutilityinterface.h]: The destructor should be virtual if we want to be able to mock this interface.

File: common/filetoolkitwithundo.cpp [hunk 9]
```diff
@@ -174,6 +174,35 @@ bool FileToolkitWithUndo::syncOverlayMount(const std::string &lower,
...
+    std::string fstype = "tmpfs";
+    unsigned long flags = 0;
+    std::unique_ptr<CreateDir> createDirInstance = std::unique_ptr<CreateDir>(new CreateDir());
+
+    if (!createDirInstance->createDirectory(dst)) {
+        log_error() << "Failed to create " << dst << " directory for tmpfs mount";
+        return false;
+    }
+
+    std::string mountoptions = logging::StringBuilder() << "size=" << maxSize;
```
Comment (tobsan) [inline common/filetoolkitwithundo.cpp:188]: This can be done with std::to_string as well I think. No need to change now, just a thought.

File: libsoftwarecontainer/src/softwarecontainer.cpp [hunk 10]
```diff
@@ -61,18 +61,25 @@ SoftwareContainer::SoftwareContainer(const ContainerID id,
     m_config(std::move(config)),
     m_previouslyConfigured(false)
 {
-    checkWorkspace();
+    m_containerRoot = buildPath(m_config->sharedMountsDir(), "SC-" + std::to_string(id));
+    checkContainerRoot(m_containerRoot);
+    if (m_config->enableWriteBuffer()) {
+        tmpfsMount(m_containerRoot, 100485760);
```
Comment (tobsan) [inline libsoftwarecontainer/src/softwarecontainer.cpp]: Magic number.

**Review by tobsan — CHANGES_REQUESTED** [review]

File: libsoftwarecontainer/src/softwarecontainer.cpp (hunk 10 above)
Comment (frznlogic) [reply libsoftwarecontainer/src/softwarecontainer.cpp]: The next story is about removing the magic number ;).

File: agent/src/containerutilityinterface.cpp (hunk 7 above)
Comment (frznlogic) [reply agent/src/containerutilityinterface.cpp]: Copy paste from Kursats comment before, it's just a syntax difference in this case I think. It does look nicer than the original way of writing it however.

File: agent/src/containerutilityinterface.h (hunk 8 above)
Comment (frznlogic) [reply agent/src/containerutilityinterface.h]: Ah, ok... Changing it back

File: common/filetoolkitwithundo.cpp (hunk 9 above)
Comment (frznlogic) [reply common/filetoolkitwithundo.cpp:188]: Thanks;)

File: libsoftwarecontainer/src/softwarecontainer.cpp (hunk 10 above)
Comment (tobsan) [reply libsoftwarecontainer/src/softwarecontainer.cpp]: I'm a bit reluctant on merging things that are super static. Can't we at least have a variable holding the number or something so that it is clear that it shouldn't be static?

Comment (frznlogic) [reply libsoftwarecontainer/src/softwarecontainer.cpp]: Fair enough. Fixed in new updates ;).

**Review by tobsan — APPROVED** [review]

**Review by kursatkobya — APPROVED** [review]

File: agent/src/containerutilityinterface.cpp (hunk 7 above)
Comment (kursatkobya) [reply agent/src/containerutilityinterface.cpp]: it is also safer in small aspect. one can mistakenly use () and create a function instead of running constructor of variable i thought.

File: agent/src/main.cpp [hunk 11]
```diff
@@ -216,10 +216,8 @@ int main(int argc, char **argv)
         std::shared_ptr<Config> config = std::make_shared<Config>(std::move(configSources),
                                                                   ConfigDefinition::mandatory(),
                                                                   ConfigDependencies());
-        std::shared_ptr<SoftwareContainerFactory> factory =
-            std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-        std::shared_ptr<ContainerUtilityInterface> utility =
-            std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+        auto factory = std::make_shared<SoftwareContainerFactory> ();
```
Comment (breakreturn) [inline agent/src/main.cpp]: There shouldn't be spaces before the parenthesis.

File: agent/unit-test/softwarecontaineragent_unittest.cpp [hunk 12]
```diff
@@ -152,7 +154,7 @@ class SoftwareContainerAgentTest: public ::testing::Test
         Glib::RefPtr<Glib::MainContext> mainContext = Glib::MainContext::get_default();
         testContainerInterface = std::shared_ptr<::testing::NiceMock<TestContainerInterface>> (new ::testing::NiceMock<TestContainerInterface>());
         factory = std::shared_ptr<SoftwareContainerFactory> (new TestFactory(testContainerInterface));
-        containerUtility = std::shared_ptr<::testing::NiceMock<MockUtility>> (new ::testing::NiceMock<MockUtility>());
+        containerUtility = std::shared_ptr<::testing::NiceMock<MockUtility>> (new ::testing::NiceMock<MockUtility>(config));
```
Comment (breakreturn) [inline agent/unit-test/softwarecontaineragent_unittest.cpp]: Remove the space before parenthesis.

File: libsoftwarecontainer/src/softwarecontainer.cpp [hunk 13]
```diff
@@ -61,18 +61,26 @@ SoftwareContainer::SoftwareContainer(const ContainerID id,
...
     m_container = std::shared_ptr<ContainerAbstractInterface>(
         new Container("SC-" + std::to_string(id),
                       m_config->containerConfigPath(),
-                      m_config->sharedMountsDir(),
+                      m_containerRoot,
                       m_config->enableWriteBuffer(),
                       m_config->containerShutdownTimeout()));
 
+
+
```
Comment (breakreturn) [inline libsoftwarecontainer/src/softwarecontainer.cpp]: Why all the whitespace here?

File: servicetest/filesystem/test_filesystem.py [hunk 14]
```diff
@@ -202,3 +203,56 @@ def test_bindmount_sockets(self, flag):
 
         if os.path.exists(absoluteTestFile):
             os.remove(absoluteTestFile)
+
+    @pytest.mark.xfail("platform.release() <= \"3.18.0\"")
+    @pytest.mark.parametrize("flag", [True, False])
+    def test_tmpfs_writebuffer_size(self, flag):
+        """ The SoftwareContainer will create a tmpfs in the location of the
+        containers temporary directory location if the write buffer is enabled.
```
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: Formatting of these comment are not as we usually do it, please have a look through the code for style and compare to the other modules in the service test tree.

File: servicetest/filesystem/test_filesystem.py [hunk 15]
```diff
@@ -202,3 +203,56 @@ def test_bindmount_sockets(self, flag):
...
+
+        Note: The filesystem is hard coded to 100MB at the time this test was
+        written, but will be configurable via the DATA field in a short while
+        """
+
+        absoluteTestFile = os.path.join("/tmp/container/SC-0/rootfs-upper", TESTFILE)
+
+        ca = Container()
+
+        if flag is True:
```
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: I'm not a fan of using parametrization like this. What's the benefit compared to having two tests? The code is mostly not reused.

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: libsoftwarecontainer/src/softwarecontainer.cpp (hunk 13 above)
Comment (frznlogic) [reply libsoftwarecontainer/src/softwarecontainer.cpp]: Because I like it :D (Fixed)

**Review by breakreturn — APPROVED** [review]
